쥬스 메이커 [STEP 1] Lingo, Taeangel

seu0313 opened the pull request:
@jungseungyeo 
안녕하세요 린생 😃
9_group의 Lingo, Taeangel 입니다. 잘 부탁드립니다!
쥬스 메이커 STEP 1 PR 보냅니다.

> 코딩 컨벤션은 스타일쉐어 가이드를 참고하였습니다. (탭사이즈: 2)

## UML
<img width="600" alt="스크린샷 2022-02-23 오후 8 24 03" src="https://user-images.githubusercontent.com/94151993/155310369-f20f9432-ff4e-49f7-b1f3-41b41cdcee72.png">

## 고민한 점
1. `make()` 함수의 에러 처리에서 `catch`로 받을 때 각각의 case를 모두 적어주는 것을 해결할 방법이 있을까? 를 고민했습니다.
2. 왜 `FruitStore` 는 클래스로 되어 있지만 `JuiceMaker`는 구조체로 구현되어 있을까에 대해 고민했습니다. 먼저,  클래스는 참조타입이기 때문에 인스턴스를 복사하지 않고 참조하게 되어 `JuiceMaker` 내에서 동일한 `FruitStore` 인스턴스로 접근할 수 있도록 하기 위해 사용했고 `JuiceMaker` 의 경우 그럴 필요가 없기 때문에 구조체로 정의했다고 생각합니다.
3. `FruitStore`와 `JuiceMaker` 객체의 기능을 구현할 때 재료가 충족하는 지 확인하는 `canMake()` 함수를 어디에다가 구현할 지 고민했고 `FruitStore`은 과일 재고를 관리/변경하는 역할을  `JuiceMaker`는 주스를 제조와 관련된 역할만 수행할 수 있도록 구분하여 `JuiceMaker`에 구현 했습니다.

## 해결한 점
1. `catch` 에서 받은 `error` 를 `MakeJuiceError` 타입으로 타입 캐스팅하여 처리했습니다. 타입을 체크할 수 있는 `is` 를 사용하여 개선하는 방법도 고민했지만, 출력하는 부분에도 직접 타입에 접근해야하는 문제가 있기 때문에 `let error`로 받고 `MakeJuiceError` 로 캐스팅하여 `rawValue`를 출력하도록 변경하였습니다.

```swift
// JuiceMaker.swift
// 기존 코드
do {
  for (fruit, amount) in juice.recipe {
    try self.fruitStore.changeNotCheckStock(of: fruit, by: -amount)
  }
} catch MakeJuiceError.outOfStock {
  print(MakeJuiceError.outOfStock.rawValue)
} catch MakeJuiceError.notExistFruit {
  print(MakeJuiceError.notExistFruit.rawValue)
} catch {
  print(error.localizedDescription)
}

// 해결한 코드
do {
  for (fruit, amount) in juice.recipe {
    try self.fruitStore.changeNotCheckStock(of: fruit, by: -amount)
  }
} catch let error as MakeJuiceError {
  print(error.rawValue)
} catch {
  print(error.localizedDescription)
}
```

2. 생성자 내부에서 `stock` 데이터를 추가했었습니다. 하지만, Fruit의 종류가 추가될 때마다 변경해줘야 한다고 생각했고 `CaseIterable` 프로토콜을 채택하여 `allCases` 프로퍼티로 처리하였고 `configureStock()` 함수로 분리하였습니다.

📒 **CaseIterable** 프로토콜이란?
열거형의 모든 case 값들에 대한 타입 프로퍼티 `allCases`으로 배열 컬렉션을 제공하는 프로토콜입니다. 따라서, `allCases` 는 배열 컬렉션`[Fruit]`를 반환하게 됩니다.

```swift
// Fruit.swift
enum Fruit: CaseIterable {
  case strawberry, banana, pineapple, kiwi, mango
}

// FruitStore.swift
// 기존 코드
init() {
    let defaultCount = 10
    
    self.stock = [
      .strawberry: defaultCount,
      .banana: defaultCount,
      .mango: defaultCount,
      .kiwi: defaultCount,
      .pineapple: defaultCount
    ]
  }

// 해결한 코드
private func configureStock() {
    let defaultCount = 10
    
    for fruit in Fruit.allCases {
      stock[fruit] = defaultCount
    }
  }
``` 

## 조언 받고 싶은 점
JuiceMaker의 `make()` 함수를 구현하던 중 `canMake()` 함수에서 **재고 개수가 0보다 크거나 같은 지를 검사**하는 로직이 있는데 FruitStore의 `changeStock()` 함수에서도 체크를 하기 때문에 중복되는 로직이 아닌가? 라는 이야기가 나왔습니다. 

로직이 중복되기 때문에 `changeStock()` 함수를 2개로 분리하여 하나는 체크하는 로직을 넣고, 다른 하나는 체크하지 않도록 해서 `canMake()` 에서만 검사하도록 하는 `changeNotCheckStock()` 함수를 새로 만들자는 이야기가 나왔는 데 어떻게 하는 것이 좋은 방법인지 궁금합니다.

[# 질문 사항 커밋 링크](https://github.com/llingo/ios-juice-maker/commit/76cb3f444d569ae82147fc1396931ea8622e91be)

```swift
// FruitStore.swift
// 기존 함수
func changeStock(of fruit: Fruit, by amount: Int) {
  guard let item = self.stock[fruit] else {
    return
  }
  guard item + amount >= 0 else {
    return
  }
  self.stock[fruit] = item + amount
}

// 분리된 함수
func changeNotCheckStock(of fruit: Fruit, by amount: Int) {
  guard let item = self.stock[fruit] else {
    return
  }
  self.stock[fruit] = item + amount
}
```

```swift
// JuiceMaker.swift
private func canMake(_ juice: Juice) -> Bool {
  for (fruit, amount) in juice.recipe {
    guard let item = self.fruitStore.stock[fruit], item >= amount else {
      return false
    }
  }
  return true
}

func make(_ juice: Juice) {
  guard self.canMake(juice) else {
    return
  }
  do {
    for (fruit, amount) in juice.recipe {
      try self.fruitStore.changeNotCheckStock(of: fruit, by: -amount)
    }
  } catch is MakeJuiceError {
    
  } catch {
    print(error.localizedDescription)
  }
}
```

Commit:
feat: Fruit, Juice 열거형 타입 생성

과일과 쥬스를 구분할 열거형 타입을 각각 생성
코딩컨벤션(스타일쉐어 가이드)에 따라 탭 사이즈 2 로 변경

Taeangel committed:
feat: FruitStore 생성

과일데이터를 dictionary로 구현하였고 각 과일의 수량을 변경하는 함수생성

Commit:
docs: README 파일 추가

Taeangel committed:
feat: configureStock, addStock 함수 생성

기존 초기화 하는 구문을 configureStock 함수로 분리했고 change에서 addStock로
함수명을 변경하고 guard문으로  바인딩을 하여 변화량 만큼 더해주도록 구현

Commit:
feat: Juice 열거형 타입에 recipe 연산 프로퍼티 생성

switch 문을 사용하여 딕셔너리에 쥬스 제작에 필요한 과일과 개수를 담은 레시피를 생성

Taeangel committed:
feat: JuiceMaker의  canMake함수와 make함수 생성

재료가 충족되는지 확인하는 canMake함수와 쥬스를 만드는 make함수를 구현
FruitStore 질문사항 추가

Commit:
feat: 에러 핸들링 구현

MakeJuiceError 에러 타입을 정의하여 재고가 부족한 경우, 존재하지 않는 과일인 경우를 case 로 설정하여 enum 타입으로 관리
FruitStore의 각각의 과일의 수를 변경하는 함수와 JuiceMaker의 make 함수에 에러 처리를 추가

Commit:
refactor: JuiceMaker의 make 함수 부분의 누락된 에러처리 변경

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,40 @@
 
-// 과일 저장소 타입
-class FruitStore {
+final class FruitStore {
```
final을 사용한 이유가 있나요?

seu0313 replied:
`FruitStore` 클래스를 다른 클래스에 상속할 필요가 없기 때문에 `final`을 사용하여 재정의를 막기 위함입니다. 만약, 재정의를 하려고 하면 컴파일 시 에러가 발생하여 실수를 방지할 수 있습니다.

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -0,0 +1,32 @@
+  case strawberry, banana, pineapple, kiwi, mango
+  case strawberryBanana, mangoKiwi
+  
+  var recipe: [Fruit: Int] {
```
너무 좋습니다

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,35 @@
-    
+  private let fruitStore = FruitStore()
+
+  private func canMake(_ juice: Juice) -> Bool {
```
이렇게 체크하는 함수를 만들고 private으로 감춘거 너무 좋아요 :)

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,35 @@
+  }
+  
+  func make(_ juice: Juice) {
+    guard self.canMake(juice) else {
```
여기서 의문이 드는거 같아요. canMake를 통과하였는데 밑에서 만들수 없는 경우가 존재하는 건가요? CanMake를 함수를 통과했는데 throw를 하는 경우가 있다면 canMake 함수를 신뢰 할 수 있나요?

seu0313 replied:
`canMake()` 함수에서 발생 가능한 에러를 모두 처리해주고 `changeStock()`는 에러 처리 없이 실행하도록 수정했습니다.

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,35 @@
+        try self.fruitStore.changeNotCheckStock(of: fruit, by: -amount)
+      }
+    } catch let error as MakeJuiceError {
+      print(error.rawValue)
```
rawValue보다는 에러 메시지를 노출하는 것이기 때문에 다른 방법이 있을거같아요. LocalizedError 이것을 찾아보고 적용해보면 좋을거 같아요

seu0313 replied:
`LocalizedError` 프로토콜을 채택하여 수정했습니다. 리팩토링 과정에서 반환 타입이 `Result` 로 변경되면서 에러 메시지를 노출하는 출력문을 제거하였습니다.

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,40 @@
+  private func configureStock() {
+    let defaultCount = 10
     
+    for fruit in Fruit.allCases {
```
이 부분 너무 좋네요 CaseIterable를 채택하고 for문으로 넣어주는 부분이 너무 좋았습니다.

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,40 @@
+    guard let item = self.stock[fruit] else {
+      throw MakeJuiceError.notExistFruit
+    }
+    self.stock[fruit] = item + amount
```
위의 두 함수의 역할이 비슷한거 같아요. 중간에 갯수를 체크하는 부분이 다르지만 사실상 같은 역할의 코드로 보여요. 하나로 코드로 만들어서 사용하는 방법이 좋을거같아요.

seu0313 replied:
1. `JuiceMaker`의 `canMake()`함수와 `FruitStore`의 `changeStock()` 함수의 체크하는 로직이 중복되기 때문에 `checkStock()` 함수를 생성하여 중복을 제거했습니다.
```swift
func checkStock(
  of fruit: Fruit,
  in amount: Int,
  by operation: (Int, Int) -> Int
) -> Result<Int, MakeJuiceError> {
  guard let item = self.stock[fruit] else {
    return .failure(.notExistFruit)
  }
  guard operation(item, amount) >= 0 else {
    return .failure(.outOfStock)
  }
  return .success(item)
}
```

2. `changeStock()`와 `changeNotCheckStock()` 함수가 stock의 값을 바꾼다는 점이 중복되어 하나로 합쳤습니다.
```swift
func changeStock(of fruit: Fruit, in amount: Int, by operation: (Int, Int) -> Int) {
  guard let item = self.stock[fruit] else {
    return
  }
  self.stock[fruit] = operation(item, amount)
}
```

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,35 @@
+    }
+    do {
+      for (fruit, amount) in juice.recipe {
+        try self.fruitStore.changeNotCheckStock(of: fruit, by: -amount)
```
내부 함수에서는 +를 해주고 있는데 여기서 -를 넣어주면 코드를 읽는데 있어서 혼란이 있을거같아요. 분명 함수에서는 재료를 더해주는 것 처럼 이니깐요.

seu0313 replied:
클로저를 사용하여 혼란이 될 + / - 를 외부에서 정의하도록 수정했습니다.
외부에서 받을 클로저가 `(Int, Int) -> Int` 타입이므로 연산자 메소드를 사용할 수 있도록 하여 혼란을 줄여보았습니다.

```swift
// 구현부
func changeStock(of fruit: Fruit, in amount: Int, by operation: (Int, Int) -> Int) {
  guard let item = self.stock[fruit] else {
    return
  }
  self.stock[fruit] = operation(item, amount)
}

// 사용
self.fruitStore.changeStock(of: fruit, in: amount, by: +)
self.fruitStore.changeStock(of: fruit, in: amount, by: -)
```

jungseungyeo replied:
제 의도는 연산자까지 주입을 해주는 것이 아니라 무엇가를 시키는 것을 원하는 거였습니다. 외부에서 어떻게 어떤 연산을 해라 이게 아니라 
consumeStock(fruit: Fruit, amount: Int)와 같이 변화라와는 다른 느낌의 네이밍과 코드를 의도하였습니다. 

외부에서 어떻게 연산 될 지를 고려해서 코드를 작성하다보면 점점 외부에 로직이 노출되어 의존관계가 형성 될 수 있습니다.

jungseungyeo reviewed (commented):
@llingo , @Taeangel  너무 고생하셨어요 코멘트를 보고 답변 달아주시면 좋을거같아요.

고민한 점과 해결한 점의 대해서 읽어보았는데요. 너무 잘 해결하신거같아요. 
기존코드와 해결한 코드의 보여주면서 무엇을 고민했는지 바로 이해가 되어서 너무 좋았습니다.

### 질문

1. canMake와 changeStock 중복인가?

`제가 보기에는 내부적으로 체크하는 부분이 있어 체크로직은 중복이라고 말을 할 수 있을거같아요. 하나의 함수가 하나의 역할을 해야 된다는 기준을 가지고 하시면 어느 부분이 문제인지 알 수 있을거같아요. `

2. changeStock와 changeNotCheckStock의 함수 로직을 다르게 하였는데 어떤 좋은 방법이 있나요?

`개인적으로 두 함수의 역할은 같다고 생각이 들어요. 같은 이유는 두 함수가 존재하는 목적이 같다는 생각이였습니다. stock이라는 딕션너리의 갯수를 변경하는 역할을 둘다 하고 있어 사실상 같은 역할이라고 판단하였고 하나는 지워도 괜찮다고 생각하고 있습니다.`

두분다 코드를 보면서 에러처리와 어느 포인트에서 고민을 하셨는지 알 수 있었던거 같아요. 아쉬운 점은 하나의 함수가 하나의 역할을 해야 하는데 몇개 그러지 못한 부분이 있었습니다. 

솔리드 원칙중 단일 책인 원칙이라는 말이 있습니다. 하나의 책임만을 가져야 한다는 말입니다. 함수 내부에서 여러가지를 할려다 보니 중복 코드가 생기고 다양한 형태가 생성 되는거 같아요. 

단일 책임을 생각하여 함수를 수정해보면 좋을거같아요. 
- changeNotCheckStock
- changeStock
수정해보면 좋을거같습니다. :) 

궁금한 점 있으면 언제든지 물어봐주세요 :)

Taeangel committed:
refactor: 중복된 로직과 함수 수정

중복된 로직을 수정하면서 FruitStore의 stock가 더이상 private(set)일 필요가 없기 때문에 private로 변경
changeNotCheckStock과 changeStock canMake 함수가 중복되었기 때문에  checkStock함수를 만들어 중복을 제거
연산할 로직을 클로저로 받도록 변경

Commit:
refactor: LocalizedError 프로토콜을 채택하여 에러 처리

MakeJuiceError의 rawValue 타입을 제거하고 extension 로 확장하여 LocalizedError 프로토콜을 채택하여 에러 처리

seu0313 commented:
리팩토링을 진행하다가 궁금한 점이 생겼습니다.
1. 함수의 반환 값을 사용하지 않을 때 `_` 를 사용해도 되는 지 궁금합니다. (기능  상으로는 가능하지만 코드의 가독성을 해치지 않을까? 라는 고민이 있어 실제로 사용해도 되는 지 궁금합니다.)
```swift
private func canMake(_ juice: Juice) -> Bool {
  for (fruit, amount) in juice.recipe {
    guard let _ = try? self.fruitStore.checkStock(of: fruit, in: amount, by: -).get() else {
      return false
    }
  }
  return true
}
```

2. FruitStore 의 `checkStock()` 에서 발생한 에러를 JuiceMaker 의 `canMake()`함수에서 받고 다시 에러를 `make()` 함수로 받아
do-catch 를 통해 처리할 때, `checkStock()` -> `canMake()` -> `make()` 로 연달아 에러가 넘겨지도록 하면 결합도가 증가하게 되는 지 궁금합니다.

Commit:
fix: checkStock 함수 오류 수정

checkStock 함수의 재고 차감 부분에서 0보다 작은 값에 대해 에러 처리하지 못했던 오류를 수정

jungseungyeo commented:
1. 함수의 반환 값을 사용하지 않을 때 _ 를 사용해도 되는 지 궁금합니다. (기능 상으로는 가능하지만 코드의 가독성을 해치지 않을까? 라는 고민이 있어 실제로 사용해도 되는 지 궁금합니다.)

``` swift
guard case let .success(item) = fruitStore.checkStock(of: fruit, in: amount, by: -) else {
    return false
}
```

이런 형태로 많이 사용해요 ㅎㅎ

2. FruitStore 의 checkStock() 에서 발생한 에러를 JuiceMaker 의 canMake()함수에서 받고 다시 에러를 make() 함수로 받아
do-catch 를 통해 처리할 때, checkStock() -> canMake() -> make() 로 연달아 에러가 넘겨지도록 하면 결합도가 증가하게 되는 지 궁금합니다.

`넵 당연히 결합도가 증가합니다. 왜 결합도가 높아졌을까요? checkStock이라는 함수의 존재의 대해서 생각해 볼 필요가 있어요. 이게  객체가 다를 객체에서 너 상태가 어떠니? 라고 물어본다면 당연히 결합도는 높아질수 밖에 없습니다. 지금 제고를 체크해줘 라고 물어보고 있고 당연히 물어본 결과의 값의 의해 분기문이 만들어지고 이러다보면 결합도는 높아지게 됩니다. ㅜㅜ`

1번째 질문의 코드는 좋지 않아 보여요 하지만 머지를 할 정도는 아니고 학습을 목표로한 step1은 잘 이해하고 있는 것으로 판단이 되어 머지를 진행합니다. 

다음 step에서 1번의 문제와 더 나아가 2번의 문제를 해결하면 좋을거같아요 :)
